### PR TITLE
Add querydoctor functionality

### PR DESCRIPTION
**Pull request by Alteqa**

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/QueryDoctorCommand.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        return predicate.equals(otherQueryDoctorCommand.predicate);
+    }
+
+    @Override
+    public String toString() {
+        return new ToStringBuilder(this)
+                .add("predicate", predicate)
+                .toString();
+    }
+}
```
Comment (alfaloo) [inline src/main/java/seedu/address/logic/commands/QueryDoctorCommand.java:59]: clean file, looks good to me

File: src/main/java/seedu/address/logic/commands/QueryPatientCommand.java [hunk 2]
```diff
@@ -56,4 +56,5 @@ public String toString() {
                 .add("predicate", predicate)
                 .toString();
     }
+
```
Comment (alfaloo) [inline src/main/java/seedu/address/logic/commands/QueryPatientCommand.java:59]: why did you add a blank line here?

File: src/main/java/seedu/address/logic/parser/QueryDoctorCommandParser.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            throw new ParseException(
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, QueryDoctorCommand.MESSAGE_USAGE));
+        }
+
+        String[] nameKeywords = trimmedArgs.split("\\s+");
+
+        return new QueryDoctorCommand(new DoctorNameContainsKeywordsPredicate(Arrays.asList(nameKeywords)));
+    }
+
+}
```
Comment (alfaloo) [inline src/main/java/seedu/address/logic/parser/QueryDoctorCommandParser.java:33]: this looks clean

File: src/main/java/seedu/address/model/person/DoctorNameContainsKeywordsPredicate.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+        DoctorNameContainsKeywordsPredicate otherPredicate = (DoctorNameContainsKeywordsPredicate) other;
+        return keywords.equals(otherPredicate.keywords);
+    }
+
+    @Override
+    public String toString() {
+        return new ToStringBuilder(this).add("keywords", keywords).toString();
+    }
+}
```
Comment (alfaloo) [inline src/main/java/seedu/address/model/person/DoctorNameContainsKeywordsPredicate.java:48]: looks good to me

File: src/test/java/seedu/address/logic/commands/QueryDoctorCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
+package seedu.address.logic.commands;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static seedu.address.logic.Messages.MESSAGE_PERSONS_LISTED_OVERVIEW;
+import static seedu.address.logic.commands.CommandTestUtil.assertCommandSuccess;
+import static seedu.address.testutil.TypicalPersons.DON;
+import static seedu.address.testutil.TypicalPersons.JAMES;
```
Comment (alfaloo) [inline src/test/java/seedu/address/logic/commands/QueryDoctorCommandTest.java:9]: For the  TypicalPersons, it'd be nice if we kept the alphabetical naming conventions... so after Don, we should come up with someone who's name starts with E, instead of James...

File: src/test/java/seedu/address/logic/commands/QueryDoctorCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+import static seedu.address.testutil.TypicalPersons.JAMES_GOH;
+import static seedu.address.testutil.TypicalPersons.RON;
```
Comment (alfaloo) [inline src/test/java/seedu/address/logic/commands/QueryDoctorCommandTest.java:11]: Please follow TypicalPersons naming convention

File: src/test/java/seedu/address/model/person/DoctorNameContainsKeywordsPredicateTest.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    @Test
+    public void toStringMethod() {
+        List<String> keywords = List.of("keyword1", "keyword2");
+        DoctorNameContainsKeywordsPredicate predicate = new DoctorNameContainsKeywordsPredicate(keywords);
+
+        String expected = DoctorNameContainsKeywordsPredicate.class.getCanonicalName()
+                + "{keywords=" + keywords + "}";
+        assertEquals(expected, predicate.toString());
+    }
+}
```
Comment (alfaloo) [inline src/test/java/seedu/address/model/person/DoctorNameContainsKeywordsPredicateTest.java:106]: looks like solid tests

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 8]
```diff
@@ -29,24 +29,26 @@ public class TypicalPersons {
 
     public static final Person CARL = new PatientBuilder().withNric("S2378593A").withName("Carl Sim")
             .withDoB("2005-01-03").withPhone("87436749").build();
-
-    // PATIENTS
-
```
Comment (alfaloo) [inline src/test/java/seedu/address/testutil/TypicalPersons.java:34]: why did you delete the patients comment annotation?

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 9]
```diff
@@ -67,9 +69,13 @@ public static List<Person> getTypicalPersons() {
         return new ArrayList<>(Arrays.asList(
                 ALICE,
                 BENSON,
-                CARL,
                 BOB,
-                CARL_GOH
+                CARL,
+                CARL_GOH,
+                DON,
+                JAMES,
```
Comment (alfaloo) [inline src/test/java/seedu/address/testutil/TypicalPersons.java:76]: Can we please follow naming conventions... keep it to alphabetical... why J all of a sudden?

**Review by alfaloo — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/address/logic/commands/QueryDoctorCommand.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+public class QueryDoctorCommand extends Command {
+
+    public static final String COMMAND_WORD = "querydoctor";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all doctors whose names contain any of "
+            + "the specified keywords (case-insensitive) and displays them as a list with index numbers.\n"
+            + "Parameters: KEYWORD [MORE_KEYWORDS]...\n"
+            + "Example: " + COMMAND_WORD + " alice bob charlie";
+
+    private final DoctorNameContainsKeywordsPredicate predicate;
```
Comment (officialchengyud) [inline src/main/java/seedu/address/logic/commands/QueryDoctorCommand.java:24]: Perhaps using a variable name like doctorNamePredicate would be more descriptive/clearer than just a predicate? Just something to think about.

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 11]
```diff
@@ -29,24 +29,26 @@ public class TypicalPersons {
...
-    // PATIENTS
-
     public static final Person AMY = new PatientBuilder().withNric(VALID_NRIC_AMY).withName(VALID_NAME_AMY)
             .withDoB(VALID_DOB_AMY).withPhone(VALID_PHONE_AMY).build();
     public static final Person BOB = new PatientBuilder().withNric(VALID_NRIC_BOB).withName(VALID_NAME_BOB)
             .withDoB(VALID_DOB_BOB).withPhone(VALID_PHONE_BOB).build();
     public static final Person CARL_GOH = new PatientBuilder().withNric("S1234343B").withName("Carl Goh")
             .withDoB("2001-02-24").withPhone("98173241").build();
 
+
```
Comment (officialchengyud) [inline src/test/java/seedu/address/testutil/TypicalPersons.java:39]: Is there a reason for this extra line?

File: src/main/java/seedu/address/model/person/DoctorNameContainsKeywordsPredicate.java
Comment (officialchengyud) [inline src/main/java/seedu/address/model/person/DoctorNameContainsKeywordsPredicate.java:1]: I like how clean this file of code is! Nice.

**Review by officialchengyud — COMMENTED** [review]
Just some nitpicks. Looking good, but could take a look at @alfaloo's comments - I agree with them. Nice job Eugene.

**Review by alfaloo — APPROVED** [review]

File: src/test/java/seedu/address/testutil/TypicalPersons.java (hunk 8 above)
Comment (Alteqa) [reply src/test/java/seedu/address/testutil/TypicalPersons.java:34]: When I was resolving conflicts I made a mistake.

**Review by Alteqa — COMMENTED** [review]
Resolved all comments.

File: src/test/java/seedu/address/testutil/TypicalPersons.java (hunk 9 above)
Comment (Alteqa) [reply src/test/java/seedu/address/testutil/TypicalPersons.java]: The James was not implemented by me.

File: src/test/java/seedu/address/logic/parser/AddressBookParserTest.java [hunk 12]
```diff
@@ -79,6 +91,15 @@ public void parseCommand_find() throws Exception {
         assertEquals(new FindCommand(new NameContainsKeywordsPredicate(keywords)), command);
     }
 
+    @Test
+    public void parseCommand_querydoctor() throws Exception {
+        List<String> keywords = Arrays.asList("foo", "bar", "baz");
+        QueryDoctorCommand command = (QueryDoctorCommand) parser.parseCommand(
```
Comment (ararchch) [inline src/test/java/seedu/address/logic/parser/AddressBookParserTest.java:97]: Maybe we could split this up into two lines so its easier to read?

File: src/main/java/seedu/address/logic/commands/QueryDoctorCommand.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+    public QueryDoctorCommand(DoctorNameContainsKeywordsPredicate predicate) {
+        this.predicate = predicate;
+    }
+
+    @Override
+    public CommandResult execute(Model model) {
```
Comment (ararchch) [inline src/main/java/seedu/address/logic/commands/QueryDoctorCommand.java:31]: Might be good to include a short javadocs here too

**Review by ararchch — APPROVED** [review]

**Comment by ararchch** [comment]
I think we need to resolve the merge conflict first @Alteqa, seems to be a small conflict of imports in the test file

File: src/main/java/seedu/address/model/person/DoctorNameContainsKeywordsPredicate.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
+package seedu.address.model.person;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+import seedu.address.commons.util.StringUtil;
+import seedu.address.commons.util.ToStringBuilder;
+
+/**
+ * Tests that a {@code Patient}'s {@code Name} matches any of the keywords given.
```
Comment (alfaloo) [inline src/main/java/seedu/address/model/person/DoctorNameContainsKeywordsPredicate.java:10]: Pretty sure the Patient here needs to be changed to Doctor

File: src/test/java/seedu/address/logic/commands/AddDoctorCommandTest.java [hunk 15]
```diff
@@ -5,7 +5,7 @@
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 import static seedu.address.testutil.Assert.assertThrows;
-import static seedu.address.testutil.TypicalPersons.JAMES;
+import static seedu.address.testutil.TypicalPersons.DAMES;
```
Comment (alfaloo) [inline src/test/java/seedu/address/logic/commands/AddDoctorCommandTest.java:8]: Lmao Dames... we should come up with better names in the future

**Review by alfaloo — APPROVED** [review]
